samza-core/src/main/java/org/apache/samza/task/SystemTimerSchedulerFactory.java

```diff
@@ -0,0 +1,183 @@
+   */
+  public interface TimerListener {
+    void onTimer();
+  }
+
+  private final ScheduledExecutorService executor;
+  private final Map<Object, SystemTimerScheduler> schedulers = new HashMap<>();
+  private final Map<TimerKey<?>, TimerCallback> readyTimers = new ConcurrentHashMap<>();
+  private TimerListener timerListener;
+
+  public static SystemTimerSchedulerFactory create(ScheduledExecutorService executor) {
+    return new SystemTimerSchedulerFactory(executor);
+  }
+
+  private SystemTimerSchedulerFactory(ScheduledExecutorService executor) {
+    this.executor = executor;
+  }
+
+  public <K> SystemTimerScheduler<K> getScheduler(K key) {
+    SystemTimerScheduler<K> scheduler = schedulers.get(key);
+    if (scheduler == null) {
+      scheduler = new SystemTimerScheduler<>(key);
+      schedulers.put(key, scheduler);
+    }
+    return scheduler;
+  }
+
+  public <K> void removeScheduler(K key) {
+    final SystemTimerScheduler<K> scheduler = schedulers.get(key);
+    if (scheduler != null) {
+      scheduler.cancel();
+      schedulers.remove(key);
+    }
+  }
+
+  void registerListener(TimerListener listener) {
+    timerListener = listener;
+  }
+
+  public Map<TimerKey<?>, TimerCallback> removeReadyTimers() {
+    final Map<TimerKey<?>, TimerCallback> timers = new TreeMap<>(readyTimers);
+    readyTimers.keySet().removeAll(timers.keySet());
+    return timers;
+  }
+
+  public class SystemTimerScheduler<K> {
+    private final K key;
+    private ScheduledFuture<?> scheduledFuture;
+
+    private SystemTimerScheduler(K key) {
+      this.key = key;
+    }
+
+    public void schedule(long time, TimerCallback<K> callback) {
+      final long delay = time - System.currentTimeMillis();
+      scheduledFuture = executor.schedule(() -> {
+          readyTimers.put(TimerKey.of(key, time), callback);
+
+          if (timerListener != null) {
+            timerListener.onTimer();
+          }
+        }, delay > 0 ? delay : 0, TimeUnit.MILLISECONDS);
+    }
+
+    public void cancel() {
+      if (scheduledFuture != null) {
+        scheduledFuture.cancel(false);
+      }
+    }
+  }
+
+  public static class TimerKey<K> implements Comparable<TimerKey<K>> {
+    private final K key;
+    private final long time;
+
+    static <K> TimerKey<K> of(K key, long time) {
+      return new TimerKey<>(key, time);
+    }
+
+    private TimerKey(K key, long time) {
+      this.key = key;
+      this.time = time;
+    }
+
+    public K getKey() {
+      return key;
+    }
+
+    public long getTime() {
+      return time;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+
+      TimerKey<?> timerKey = (TimerKey<?>) o;
+      if (time != ((TimerKey<?>) o).time) {
+        return false;
+      }
+      return key != null ? key.equals(timerKey.key) : timerKey.key == null;
```

Comment:
key should never be null in this class. Same for hashcode/compareTo etc.
Should add a null and uniqueness check when the key is registered.

Reply:
right, makes sense. Removed the null check.